File: katas/BowlingGame/js/src/__tests__/Game.test.ts
import { Game } from "../Game";

let game: Game;
beforeEach(() => {
  game = new Game();
});

test("gutter game should return 0 score", () => {
  rollMany(20, 0);
  expect(game.calculateScore()).toEqual(0);
});

test("all 1s should return 20", () => {
  rollMany(20, 1);
  expect(game.calculateScore()).toEqual(20);
});

test("spare adds bonus of next roll", () => {
  rollSpare();

  // Next roll after spare
  game.roll(3);

  // For rest of game, roll 0s
  rollMany(17, 0);

  // Score without bonus is 13.
  // Add bonus of 3 to 13 to get 16
  expect(game.calculateScore()).toEqual(16);
});

test("strike adds bonus of next two rolls", () => {
  rollStrike();
  game.roll(3);
  game.roll(4);
  rollMany(17, 0);

  // Score without bonus is 17
  // Score with bonus of 3 + 4 is 24
  expect(game.calculateScore()).toEqual(24);
})

test("perfect game", () => {
  rollMany(12, 10);
  expect(game.calculateScore()).toEqual(300);
})

function rollMany(numRolls: number, scorePerRoll: number) {
  for (let i = 0; i < numRolls; i++) {
    game.roll(scorePerRoll);
  }
}

function rollSpare() {
  game.roll(5);
  game.roll(5);
}

function rollStrike() {
  game.roll(10);
}

File: katas/BowlingGame/ruby/src/__tests__/Game.test.rb
require_relative "../Game"
require "test/unit"
 
class TestGame < Test::Unit::TestCase
 
  def test_gutter_game
    game = Game.new()
    roll_many(game, 20, 0)

    assert_equal(0, game.score())
  end

  def test_all_ones
    game = Game.new()
    roll_many(game, 20, 1)

    assert_equal(20, game.score())
  end

  def test_one_spare
    game = Game.new()
    roll_spare(game)
    game.roll(3)
    roll_many(game, 17, 0)
    assert_equal(16, game.score())
  end

  def test_one_strike
    game = Game.new()
    roll_strike(game)
    game.roll(3)
    game.roll(4)
    roll_many(game, 17, 0)
    assert_equal(24, game.score())
  end

  def test_perfect_game
    game = Game.new()
    roll_many(game, 12, 10)
    assert_equal(300, game.score())
  end

  def roll_many(game, num_rolls, score_for_rolls)
    for i in 1..num_rolls do
      game.roll(score_for_rolls)
    end
  end

  def roll_spare(game)
    game.roll(5)
    game.roll(5)
  end

  def roll_strike(game)
    game.roll(10)
  end
 
end
